### DISCUSSION TITLE
TypeError: Cannot set property length of #<Map> which has only a getter

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2597] [Q&A]
I want to add [`nest-mqtt`](https://github.com/128keaton/nest-mqtt) (a fork from [upstream](https://github.com/microud/nest-mqtt), see microud/nest-mqtt#10), but although it connects to the MQTT broker, MikroORM fails immediately right after the connection is made.

The error is as follows:

```ts
[Nest] 2384470  - 03/01/2022, 21:38:52   ERROR [ExceptionHandler] Cannot set property length of #<Map> which has only a getter
TypeError: Cannot set property length of #<Map> which has only a getter
    at _clone ({repo_root}/node_modules/clone/clone.js:162:16)
    at _clone ({repo_root}/node_modules/clone/clone.js:162:18)
    at clone ({repo_root}/node_modules/clone/clone.js:202:10)
    at Function.copy ({repo_root}/node_modules/@mikro-orm/core/dist/utils/Utils.js:237:36)
    at MetadataDiscovery.discoverDirectories ({repo_root}/node_modules/@mikro-orm/core/dist/metadata/MetadataDiscovery.js:137:55)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at MetadataDiscovery.findEntities ({repo_root}/node_modules/@mikro-orm/core/dist/metadata/MetadataDiscovery.js:106:9)
    at MetadataDiscovery.discover ({repo_root}/node_modules/@mikro-orm/core/dist/metadata/MetadataDiscovery.js:53:9)
    at MikroORM.discoverEntities ({repo_root}/node_modules/@mikro-orm/core/dist/MikroORM.js:87:25)
    at Function.init ({repo_root}/node_modules/@mikro-orm/core/dist/MikroORM.js:42:9)
```

I presumed the source of this error is that in the root module (where I import both `MikroOrmModule.forRoot()` and `MqttModule.forRoot()`), the class implements `NestModule` in order to work around an issue in Nest.

```ts
configure(consumer: MiddlewareConsumer) {
  consumer.apply(MikroOrmMiddleware).forRoutes('*')
}
```

However, even after removing it (and keeping `export class AppModule {}` only), the error persists.

Could you look into this please?

**Comment by tukusejssirs** [comment]
On second thought, it seems to be `clone` issue, does it? Should I report it at [their issue tracker](https://github.com/pvorb/clone/issues)?

Related PR: pvorb/clone#91

Reported here: pvorb/clone#117

**Update:** It seems like it isn’t `clone`’s fault at all.

**Reply by B4nan** [reply]
You must be doing something wrong in your entity definition, as this is coming from your metadata (so entity classes, not instances) - and you must have some object there that fails to be cloned. Usually these problems origin from resources like db connections that can't be cloned, but here it is apparently some `Map` object.

> Could you look into this please?

You haven't shared any details about your actual project, so no, there is nothing I can look into. I can only help with minimal reproductions, or I could offer paid consulting.

**Reply by tukusejssirs** [reply]
> You must be doing something wrong in your entity definition

Well, the strange thing is that without `nest-mqtt` it works fine, but then I use it, this the error is thrown. When I remove `MqttModule.forRoot()` from the module, everything works as expected.

---

> You haven't shared any details about your actual project, so no, there is nothing I can look into. I can only help with minimal reproductions, or I could offer paid consulting.

[Here](https://github.com/tukusejssirs/nest-test/tree/mikro-orm_nest-mqtt)’s a minimal reproduction. Thanks! :pray:

Tested with `npm v8.1.2` and `node v16.13.1` and `node v17.2.0`. Also note that it does not matter if the MQTT is actually running. Also change the MikroORM config in `src/mikro-orm.config.ts` as needed.

**Reply by B4nan** [reply]
I dont see any failures if I just try to start the project

![image](https://user-images.githubusercontent.com/615580/148218307-0834b298-bd14-4a33-b800-4575f6aab6a1.png)

**Reply by tukusejssirs** [reply]
Switch to `mikro-orm_nest-mqtt` branch please. :wink:

**Reply by B4nan** [reply]
That's not even possible to install on macos/m1

![image](https://user-images.githubusercontent.com/615580/148219100-eaaae2eb-4020-42e4-929b-8564e743a89a.png)

edit: ok now i see it

**Reply by tukusejssirs** [reply]
> That's not even possible to install on macos/m1

Okay, pull the latest changes please.

> edit: ok now i see it

Okay.

**Reply by B4nan** [reply]
That mqtt package needs to patch globals somehow, the problem is with cloning `EntityMetadata.propertyOrder` which is a `Map` - a regular `Map` instance, nothing special, and in case of your entity it is even empty (at least at that point where `clone` is called).

**Reply by tukusejssirs** [reply]
> That mqtt package needs to patch globals somehow, the problem is with cloning `EntityMetadata.propertyOrder` which is a `Map` - a regular `Map` instance, nothing special

I presume you mean `nest-mqtt`, not `mqtt` package (`nest-mqtt`’s dependency, which allows to create an MQTT client.

As for cloning `EntityMetadata.propertyOrder`, it is quite strange, as `nest-mqtt` should _only_ create an MQTT client, connect the server and add some decorators to pubish and subscribe messages over MQTT. See the [source code](https://github.com/128keaton/nest-mqtt/tree/master/src).

> in case of your entity it is even empty (at least at that point where `clone` is called).

Yes, because I created a _minimal_ reproduction. I get the very same error even when creating a non-empty entity and table actually exists in the DB.

**Reply by B4nan** [reply]
> I presume you mean nest-mqtt, not mqtt package (nest-mqtt’s dependency, which allows to create an MQTT client.

No, I mean `mqtt`, the problem will be probably there, not in its nest wrapper.

> Yes, because I created a minimal reproduction. I get the very same error even when creating a non-empty entity and table actually exists in the DB.

You are missing my point, your entity still has one property, which will end up in this map at some point. I said that at the point of failure, the map is still empty, that has nothing to do with how minimal your reproduction is :]

I guess this package will be the culprit, it is used in the mqtt.js and it apparently does patch the globals:

https://github.com/montagejs/collections

I can confirm that if you comment out the file in there that patches the global `Map` object, it will pass (it will fail on a `Set` as that has the very same issue). 

So to wrap up, by using the `mqtt` library, you also get the globals for `Map`, `Set`, `Array` and god knows what else are patched, and apparently in a way that the classes are not even extending the native ones, so other libraries probably fail to recognize them. In other words, due to this, `clone` does not see the type as a `Map`, but as a regular JS object, so it just iterates thru its properties and call clone on those recursively. 

This will definitely not be addressed in MikroORM anyhow.

**Reply by B4nan** [reply]
Btw the very last issue in that package is actually pointing to the `collections.js` being problematic.

https://github.com/mqttjs/MQTT.js/issues/1392

And I am sure there will be more of those.

**Reply by B4nan** [reply]
Btw if I also comment out the patching of `Set`, it will pass everything and fail just on the mqtt server not running

```
[Nest] 41282  - 05. 01. 2022 14:19:48     LOG [InstanceLoader] MikroOrmCoreModule dependencies initialized +763ms
[Nest] 41282  - 05. 01. 2022 14:19:48     LOG [RoutesResolver] AppController {/}: +5ms
[Nest] 41282  - 05. 01. 2022 14:19:48     LOG [RouterExplorer] Mapped {/, GET} route +1ms
[Nest] 41282  - 05. 01. 2022 14:19:48     LOG [MqttClient] MqttModule dependencies initialized
[Nest] 41282  - 05. 01. 2022 14:19:48     LOG [NestApplication] Nest application successfully started +2ms
[Nest] 41282  - 05. 01. 2022 14:19:49     LOG [MqttClient] MQTT reconnecting
[Nest] 41282  - 05. 01. 2022 14:19:49   ERROR [MqttClient] Error: connect ECONNREFUSED 127.0.0.1:1883
[Nest] 41282  - 05. 01. 2022 14:19:49     LOG [MqttClient] MQTT closed
[Nest] 41282  - 05. 01. 2022 14:19:50     LOG [MqttClient] MQTT reconnecting
[Nest] 41282  - 05. 01. 2022 14:19:50   ERROR [MqttClient] Error: connect ECONNREFUSED 127.0.0.1:1883
[Nest] 41282  - 05. 01. 2022 14:19:50     LOG [MqttClient] MQTT closed
[Nest] 41282  - 05. 01. 2022 14:19:51     LOG [MqttClient] MQTT reconnecting
[Nest] 41282  - 05. 01. 2022 14:19:51   ERROR [MqttClient] Error: connect ECONNREFUSED 127.0.0.1:1883
[Nest] 41282  - 05. 01. 2022 14:19:51     LOG [MqttClient] MQTT closed
[Nest] 41282  - 05. 01. 2022 14:19:52     LOG [MqttClient] MQTT reconnecting
[Nest] 41282  - 05. 01. 2022 14:19:52   ERROR [MqttClient] Error: connect ECONNREFUSED 127.0.0.1:1883
[Nest] 41282  - 05. 01. 2022 14:19:52     LOG [MqttClient] MQTT closed
[Nest] 41282  - 05. 01. 2022 14:19:53     LOG [MqttClient] MQTT reconnecting
[Nest] 41282  - 05. 01. 2022 14:19:53   ERROR [MqttClient] Error: connect ECONNREFUSED 127.0.0.1:1883
[Nest] 41282  - 05. 01. 2022 14:19:53     LOG [MqttClient] MQTT closed
[Nest] 41282  - 05. 01. 2022 14:19:54     LOG [MqttClient] MQTT reconnecting
[Nest] 41282  - 05. 01. 2022 14:19:54   ERROR [MqttClient] Error: connect ECONNREFUSED 127.0.0.1:1883
[Nest] 41282  - 05. 01. 2022 14:19:54     LOG [MqttClient] MQTT closed
```

It is about those two files:
https://github.com/montagejs/collections/blob/master/_map.js
https://github.com/montagejs/collections/blob/master/_set.js

**Reply by tukusejssirs** [reply]
Thank you very much for time and help! :pray:

You are right. Downgrading `mqtt` to `v4.2.8` (latest version without `collections` dependency) solved the issue! :+1: :1st_place_medal:
